src/karmen_frontend/src/components/printers/__tests__/webcam-stream.js

```diff
@@ -0,0 +1,104 @@
+import React from "react";
+import { render } from "@testing-library/react";
+import configureMockStore from "redux-mock-store";
+import thunk from "redux-thunk";
+import { Provider } from "react-redux";
+import WebcamStream from "../webcam-stream";
+
+require("jest-fetch-mock").enableMocks();
+const printerUuid = "20e91c14-c3e4-4fe9-a066-e69d53324a20";
+const orgUuid = "randomUuidOfOrganisation";
+
+const createPrinter = (connected) => {
+  return {
+    uuid: printerUuid,
+    client: {
+      access_level: "unlocked",
+      api_key: null,
+      connected: connected,
+      name: "octoprint",
+      pill_info: {
+        karmen_version: "0.2.0",
+        update_available: "0.2.2",
+        update_status: null,
+        version_number: "0.2.0",
+      },
+      plugins: ["awesome_karmen_led"],
+      version: {
+        api: "0.1",
+        server: "0.0.1",
+        text: "octoprint fake",
+      },
+    },
+  };
+};
+
+const middlewares = [thunk];
+const mockStore = configureMockStore(middlewares);
+
+const renderWebcamStream = (store, printer) => {
+  return render(
+    <Provider store={store}>
+      <WebcamStream
```

Comment:
Since we're doing some changes here, how difficult would be to liberate webcam stream from the underlying store for better testability? I'm thinking a simple provider would do the job easily and we wouldn't need to deal with store in tests (which also makes the test more easy to maintain long-term).

Reply:
I can imagine to remove redux connection and pass the params thru a props.

Reply:
Exactly.

Reply:
I've considered an impact of removing a connection to a redux and I found out that it would cause an useless rerendering of other components (other printers). Think about this way: each component WebcamStream runs `setInterval()` and it asynchronously loads an image and update a state (webcam image). If I would move this logic to an upper component (PrinterList) where I would get all printers' webcam image, each change of certain image will run a rerendering of other printers.

Reply:
I didn't want you to move this logic up the chain. I suggested breaking this one component into two that can be tested separately:

1) a webcam stream provider - reads data from the store and provides it for the children
2) webcam stream loader - usually wrapped by the provider but can be tested in isolation

Reply:
In the app, it would always be used in combination. In tests, we can test either separately. Makes sense?

Reply:
I guess the provided example wasn't a best one as they use contexts whereas I would simply use props. My only motivation was splitting the store retrieval and displaying the stream itself into parts so that each can be tested in isolation.

Reply:
Check the new commit, I've splitted it.

Reply:
Great!